lib/propolis/src/accessors.rs

```diff
@@ -12,54 +12,98 @@
 //! contained emulation should fail any DMA accesses.
 
 use std::collections::{BTreeMap, BTreeSet, VecDeque};
+use std::ffi::c_void;
 use std::marker::PhantomData;
+use std::ptr::NonNull;
 use std::sync::{Arc, Mutex, MutexGuard, Weak};
 
 use crate::vmm::{MemCtx, VmmHdl};
 
-type AccId = usize;
-
-const ID_NULL: AccId = 0;
-const ID_ROOT: AccId = 1;
+/// Key type for identifying nodes referenced by `Tree`.
+#[derive(Ord, PartialOrd, Eq, PartialEq, Debug, Copy, Clone)]
+pub struct NodeKey(NonNull<Node<c_void>>);
+impl<T> From<&Arc<Node<T>>> for NodeKey {
+    fn from(value: &Arc<Node<T>>) -> Self {
+        let raw = Arc::as_ptr(value) as *const Node<c_void>;
+        let inner = unsafe { NonNull::new_unchecked(raw as *mut Node<c_void>) };
+        NodeKey(inner)
+    }
+}
+impl std::fmt::Display for NodeKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:p}", self.0.as_ptr())
+    }
+}
+// Safety: While the key uses a pointer (!Send) type internally, it is for
+// unique identification purposes only, and is never meant to be dereferenced,
+// copied from, or transformed into a reference of any kind.
+unsafe impl Send for NodeKey {}
 
 struct TreeNode<T> {
-    parent_id: AccId,
+    /// [NodeKey] of the parent to this node
+    ///
+    /// Holds [None] if the node is the root of the [Tree]
+    parent_key: Option<NodeKey>,
+    /// [Weak] reference back to the node.  This access is needed if the node
+    /// undergoes adoption (being moved to a different [Tree])
     node_ref: Weak<Node<T>>,
-    children: BTreeSet<AccId>,
+    /// List of keys to child nodes (if any)
+    children: BTreeSet<NodeKey>,
+    /// Display name for [Tree::print()]-ing
     name: Option<String>,
 }
 impl<T> TreeNode<T> {
     fn new(
-        parent_id: AccId,
+        parent_key: NodeKey,
         node_ref: Weak<Node<T>>,
         name: Option<String>,
     ) -> Self {
-        Self { parent_id, node_ref, children: BTreeSet::new(), name }
+        Self {
+            parent_key: Some(parent_key),
+            node_ref,
+            children: BTreeSet::new(),
+            name,
+        }
+    }
+    fn new_root(node_ref: Weak<Node<T>>) -> Self {
+        Self {
+            parent_key: None,
+            node_ref,
+            children: BTreeSet::new(),
+            name: None,
+        }
     }
 }
+
 struct Tree<T> {
+    /// Underlying resource (if any) that this heirarchy is granting access to
     resource_root: Option<Arc<T>>,
-    nodes: BTreeMap<AccId, TreeNode<T>>,
-    next_id: AccId,
+    /// Key of the root node of this hierarchy
+    ///
+    /// Only when the tree is being initialized, should `root_key` be [None]
+    root_key: Option<NodeKey>,
+    /// Nodes within this heirarchy
```

Comment:
spelling nit:
```suggestion
    /// Nodes within this hierarchy
```

Reply:
*sigh* I don't know why I haven't enabled `:set spell` all the time.